### PR TITLE
Pydantic model

### PR DESCRIPTION
**Pull request by zain-sohail**
Please see the config model for complete possible schema. I extracted all parts from mpes and flashloader currently. Though mpes yaml config is not yet updated with the ConfigModel.

To run
```python
from sed.config.config_model import ConfigModel, CoreModel
from sed.core.config import parse_config
config_file = '../sed/config/flash_example_config.yaml'
config = parse_config(config_file)
model = CoreModel(**config["core"])
print(model)
print(model.loader)

model = ConfigModel(**config)
print(model)

config_dict_verified = model.model_dump()
```

This PR currently won't pass tests as it has config alterations that are breaking. Since creating a Pydantic schema/model, I thought it's good if the dataframe section is better streamlined.

So now there is a columns section. Instead of
```yaml
  x_column: X  
  y_column: Y
  tof_column: t
  corrected_x: Xm
  ...
```

we would need 
```yaml
  columns:
    x: X
    y: Y
    tof: t
    corrected_x: Xm
    ...
``` 

Another thing to discuss: Whether we do a dict dump after schema verification or use the model itself. I prefer the model, as it makes a nice object oriented way of accessing values.

I won't update the other codebase until we decide if these alterations make sense. A Pydantic model for the updated config is also included here. This can be easily adapted to the config schema in v1, if necessary.

@rettigl please take a look.

### CODE REVIEW

**Comment by rettigl** [comment]
Generally, I find this a good idea, and also support directly using the pydantic models as config items. However, I don't like the idea to put again so many defaults into the code. If we remove all values there, and just make all parameters that are part of the default config as required, we would also get a consistent set, and also can make sure that certain values are available in the modules, no?

**Comment by zain-sohail** [comment]
> Generally, I find this a good idea, and also support directly using the pydantic models as config items. However, I don't like the idea to put again so many defaults into the code. If we remove all values there, and just make all parameters that are part of the default config as required, we would also get a consistent set, and also can make sure that certain values are available in the modules, no?

Makes sense. I don't think there should be no problem in moving the defaults to how it currently is (in v1 branch). I did this because pydantic seems to promote the idea of defaults within model.

I am still not satisfied with how the config and model looks for channels. Will try to think of a better structure.

File: sed/config/config_model.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        datafile: FilePath  # .h5 extension in filepath
+
+    calibration: Optional[Calibration] = None
+
+    class Offsets(BaseModel):
+        creation_date: Optional[datetime] = None
+        constant: Optional[float] = None
+        flip_delay_axis: Optional[bool] = False
+
+        ## This seems rather complicated way to define offsets,
```
Comment (zain-sohail) [inline sed/config/config_model.py:220]: As mentioned here, offsets has quite a convoluted config structure, and it differs from the passable args. The add_offsets method thus mostly tries to figure out what is being passed. I'd suggest we try refactoring the method and fix a specific config model for it.
I'd probably suggest to remove the dict like column entries completely.

Comment (rettigl) [reply sed/config/config_model.py:220]: I can have a look. In principle, the structure is not so complicated. There are top level entries like constant, flip etc., and then a sub dictionary for each column with options like weights, preserve mean etc.

Comment (rettigl) [reply sed/config/config_model.py:220]: The parameters are a bit different than the config dict, that is true, but the are refined into that structure. I think your config model looks okay, not sure I understand the problem correctly.

**Comment by rettigl** [comment]
@zain-sohail  I updated the mode and code so it passes tests and has all necessary fields. There are still some checks that could be added, and some tests for the schema could be added.

File: sed/config/config_model.py (hunk 1 above)
Comment (zain-sohail) [reply sed/config/config_model.py:301]: I just found the way the value unpacking rather convoluted, and was trying to refactor it. But naturally, that is not what this PR is about.

**Comment by zain-sohail** [comment]
> @zain-sohail I updated the mode and code so it passes tests and has all necessary fields. There are still some checks that could be added, and some tests for the schema could be added.

Thanks a lot for fixing that.

What checks are you thinking of?

Regarding tests, generally, since Pydantic itself is well tested, we can expect it to do what it is supposed to. Or did you mean some integration testing? But I do find that after this config validation, a lot of checks within the rest code can be removed. Either I can look into it in this PR or we merge this and do that later.

**Comment by rettigl** [comment]
I thought about tests that the config fits to the model, and fails if additional entries are present, or required missing. The former e g was not working on your version.
Also, further checks regarding the consistency can be added to the config model. Some I can still do.
Generally, I suggest merging this soon, so we can move forward with v1.

File: sed/core/config.py [hunk 2]
```diff
@@ -141,7 +144,11 @@ def parse_config(
         base_dictionary=default_dict,
     )
 
-    return config_dict
+    if not verify_config:
+        return config_dict
+    # Run the config through the ConfigModel to ensure it is valid
+    config_model = ConfigModel(**config_dict)
+    return config_model.model_dump(exclude_unset=True, exclude_none=True, exclude_defaults=True)
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/core/config.py:151]: As defaults are only None, this is probably redundant, you are right. I will remove

File: sed/config/config_model.py [hunk 3]
```diff
@@ -51,6 +68,14 @@ def validate_loader(cls, v: str) -> str:
             raise ValueError(f"Invalid loader {v}. Available loaders are: {names}")
         return v
 
+    @field_validator("num_cores")
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:72]: That's possible. This other check we do in processor, so I see no need here.

File: sed/config/config_model.py [hunk 4]
```diff
@@ -309,7 +326,7 @@ class NexusModel(BaseModel):
     reader: str  # prob good to have validation here
     # Currently only NXmpes definition is supported
     definition: Literal["NXmpes"]
-    input_files: Sequence[FilePath]
+    input_files: Sequence[str]
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:356]: There were problems with the other settings, but I can investigate again.

File: tests/data/loader/flash/config.yaml [hunk 5]
```diff
@@ -125,32 +101,27 @@ dataframe:
     # The auxiliary channel has a special structure where the group further contains
     # a multidimensional structure so further aliases are defined below
     dldAux:
-      format: per_train
+      format: per_pulse
```
[the remark this answers is not in the document]
Comment (rettigl) [reply tests/data/loader/flash/config.yaml:104]: I copied this from the flash config file, so it also needs to change there then.

File: sed/config/config_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py]: Changed back and fixed elsewhere

File: sed/config/config_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zain-sohail) [reply sed/config/config_model.py]: I'd suggest we move all config validation to the model, instead of doing it in main code.

**Comment by zain-sohail** [comment]
Should we merge this?

**Comment by rettigl** [comment]
> Should we merge this?

There is still something wrong in the tutorials. I will have a look. I will also give it another review then.

File: tests/test_config.py [hunk 6]
```diff
@@ -36,7 +41,7 @@
 
 def test_default_config() -> None:
     """Test the config loader for the default config."""
-    config = parse_config()
+    config = parse_config(verify_config=False)
```
Comment (rettigl) [inline tests/test_config.py:44]: Why verify_config=False here?

File: tests/test_config.py [hunk 7]
```diff
@@ -133,3 +145,33 @@ def test_save_dict() -> None:
             save_config(config_dict, filename, overwrite=True)
             config = load_config(filename)
             assert "test_entry" not in config.keys()
+
+
+@pytest.mark.parametrize("config_path", config_paths)
+def test_config_model_valid(config_path) -> None:
+    """Test the config model for a valid config."""
+    config = parse_config(config_path, user_config={}, system_config={}, verify_config=True)
```
Comment (rettigl) [inline tests/test_config.py:153]:
```suggestion
    config = parse_config(config_path, folder_config={}, user_config={}, system_config={}, verify_config=True)
```

File: tests/test_config.py [hunk 8]
```diff
@@ -133,3 +145,33 @@ def test_save_dict() -> None:
...
+    assert config is not None
+
+
+def test_invalid_config_extra_field():
+    """Test that an invalid config with an extra field fails validation."""
+    default_config = parse_config(user_config={}, system_config={}, verify_config=True)
```
Comment (rettigl) [inline tests/test_config.py:159]:
```suggestion
    default_config = parse_config(folder_config={}, user_config={}, system_config={}, verify_config=True)
```

File: tests/test_config.py [hunk 9]
```diff
@@ -133,3 +145,33 @@ def test_save_dict() -> None:
...
+    invalid_config = default_config.copy()
+    invalid_config["extra_field"] = "extra_value"
+    with pytest.raises(ValidationError):
+        parse_config(invalid_config, user_config={}, system_config={}, verify_config=True)
+
+
+def test_invalid_config_missing_field():
+    """Test that an invalid config with a missing required field fails validation."""
+    default_config = parse_config(user_config={}, system_config={}, verify_config=True)
```
Comment (rettigl) [inline tests/test_config.py:168]:
```suggestion
    default_config = parse_config(folder_config={}, user_config={}, system_config={}, verify_config=True)
```

File: tests/test_config.py [hunk 10]
```diff
@@ -133,3 +145,33 @@ def test_save_dict() -> None:
...
+    invalid_config = default_config.copy()
+    del invalid_config["core"]["loader"]
+    with pytest.raises(ValidationError):
+        parse_config(
+            user_config={},
```
Comment (rettigl) [inline tests/test_config.py:173]:
```suggestion
            folder_config={},
            user_config={},
```

File: tests/test_config.py [hunk 11]
```diff
@@ -133,3 +145,33 @@ def test_save_dict() -> None:
...
+    assert config is not None
+
+
+def test_invalid_config_extra_field():
+    """Test that an invalid config with an extra field fails validation."""
+    default_config = parse_config(user_config={}, system_config={}, verify_config=True)
+    invalid_config = default_config.copy()
+    invalid_config["extra_field"] = "extra_value"
+    with pytest.raises(ValidationError):
+        parse_config(invalid_config, user_config={}, system_config={}, verify_config=True)
```
Comment (rettigl) [inline tests/test_config.py:163]:
```suggestion
        parse_config(invalid_config, folder_config={}, user_config={}, system_config={}, verify_config=True)
```

File: tests/test_config.py
Comment (rettigl) [inline tests/test_config.py:1]: We could still add tests for the config validators (invalid gid, loader name)

File: tutorial/5_sxp_workflow.ipynb
Comment (rettigl) [inline tutorial/5_sxp_workflow.ipynb:1]: needs to be tested.

**Review by rettigl — APPROVED** [review]
Looks mostly good, apart from a few things still to fix. Let's resolve each point once addressed.

File: sed/config/config_model.py [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+    # write validator for model so that x_column gets converted to columns: x
+
+
+class BinningModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    hist_mode: Literal["numpy", "numba"]
+    mode: Literal["fast", "lean", "legacy"]
+    pbar: bool
+    threads_per_worker: int
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:151]: done

File: sed/config/config_model.py [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    stream_name_prefixes: Optional[dict] = None
+    stream_name_postfixes: Optional[dict] = None
+    beamtime_dir: Optional[dict] = None
+
+
+class EnergyModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    bins: int
+    ranges: Sequence[int]
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:178]: done

File: sed/config/config_model.py [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+
+    stream_name_prefixes: Optional[dict] = None
+    stream_name_postfixes: Optional[dict] = None
+    beamtime_dir: Optional[dict] = None
+
+
+class EnergyModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    bins: int
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:177]: done

File: sed/config/config_model.py [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+    mode: Literal["fast", "lean", "legacy"]
+    pbar: bool
+    threads_per_worker: int
+    threadpool_API: Literal["blas", "openmp"]
+
+
+class HistogramModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    bins: Sequence[int]
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:158]: done

File: sed/config/config_model.py [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+    first_event_time_stamp_key: Optional[str] = None
+    ms_markers_key: Optional[str] = None
+    # flash specific settings
+    forward_fill_iterations: Optional[int] = None
+    ubid_offset: Optional[int] = None
+    split_sector_id_from_dld_time: Optional[bool] = None
+    sector_id_reserved_bits: Optional[int] = None
+    sector_delays: Optional[Sequence[float]] = None
+    daq: Optional[str] = None
+    num_trains: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:140]: done

File: sed/config/config_model.py [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        amplitude2: Optional[float] = None
+
+    correction: Optional[EnergyCorrectionModel] = None
+
+
+class MomentumModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    axes: Sequence[str]
+    bins: Sequence[int]
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:241]: done

File: sed/config/config_model.py [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+        cstep: float
+
+    calibration: Optional[MomentumCalibrationModel] = None
+
+    class MomentumCorrectionModel(BaseModel):
+        model_config = ConfigDict(extra="forbid")
+
+        creation_date: Optional[datetime] = None
+        feature_points: Sequence[tuple[float, float]]
+        rotation_symmetry: int
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:269]: done

File: sed/config/config_model.py [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+    class MomentumTransformationModel(BaseModel):
+        model_config = ConfigDict(extra="forbid")
+
+        creation_date: Optional[datetime] = None
+        scale: Optional[float] = None
+        angle: Optional[float] = None
+        xtrans: Optional[float] = None
+        ytrans: Optional[float] = None
+
+    transformations: Optional[MomentumTransformationModel] = None
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:287]: done

File: sed/config/config_model.py [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+
+class DelayModel(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    adc_range: tuple[int, int]
+    flip_time_axis: bool
+    # Group keys in the datafile
+    p1_key: Optional[str] = None
+    p2_key: Optional[str] = None
+    p3_key: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/config_model.py:298]: done

File: sed/config/flash_example_config.yaml [hunk 21]
```diff
@@ -1,9 +1,8 @@
 # This file contains the default configuration for the flash loader.
-
 core:
   # defines the loader
-  loader: flash
-  # Since this will run on maxwell most probably, we have a lot of cores at our disposal
+  loader: 'flash'
+  # # Since this will run on maxwell most probably, we have a lot of cores at our disposal
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/flash_example_config.yaml:5]: done

File: sed/config/flash_example_config.yaml [hunk 22]
```diff
@@ -215,8 +195,20 @@ dataframe:
...
       dataset_key: "/zraw/FLASH.SYNC/LASER.LOCK.EXP/F1.PG.OSC/FMC0.MD22.1.ENCODER_POSITION.RD/dGroup/value"
 
+# metadata collection from scicat
+# metadata:
+#   scicat_url: <URL>
+#   scicat_token: <TOKEN>
+
+# The nexus collection routine shall be finalized soon for both instruments
+# nexus:
+#   reader: "flash"
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/config/flash_example_config.yaml:205]: done

File: sed/core/config.py [hunk 23]
```diff
@@ -55,12 +57,13 @@ def parse_config(
...
             Defaults to *package_dir*/config/default.yaml".
         verbose (bool, optional): Option to report loaded config files. Defaults to True.
+        verify_config (bool, optional): Option to verify config file. Defaults to True.
     Raises:
         TypeError: Raised if the provided file is neither *json* nor *yaml*.
         FileNotFoundError: Raised if the provided file is not found.
 
     Returns:
-        dict: Loaded and possibly completed config dictionary.
+        ConfigModel: Loaded and possibly completed pydantic config model.
```
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/core/config.py:66]: done

File: sed/loader/sxp/loader.py
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/loader/sxp/loader.py:1]: done

File: tests/data/loader/flash/config.yaml [hunk 24]
```diff
@@ -115,28 +105,23 @@ dataframe:
...
       slice: 4
-      subChannels:
+      sub_channels:
         sampleBias:
           slice: 0
-          dtype: float64
+          dtype: float32
         tofVoltage:
           slice: 1
           dtype: float64
```
[the remark this answers is not in the document]
Comment (rettigl) [reply tests/data/loader/flash/config.yaml:114]: done

File: tests/test_config.py (hunk 6 above)
Comment (rettigl) [reply tests/test_config.py:44]: replaced by config={}, folder_config={}, user_config={}, system_config={}

File: tests/test_config.py (hunk 7 above)
Comment (rettigl) [reply tests/test_config.py:153]: done

File: tests/test_config.py (hunk 8 above)
Comment (rettigl) [reply tests/test_config.py:159]: done

File: tests/test_config.py (hunk 10 above)
Comment (rettigl) [reply tests/test_config.py:173]: done

File: tests/test_config.py (hunk 11 above)
Comment (rettigl) [reply tests/test_config.py:163]: done

File: tests/test_config.py
Comment (rettigl) [reply tests/test_config.py:1]: added

File: tutorial/5_sxp_workflow.ipynb
Comment (rettigl) [reply tutorial/5_sxp_workflow.ipynb:1]: fixed and tested

File: sed/core/processor.py
[the remark this answers is not in the document]
Comment (rettigl) [reply sed/core/processor.py:1]: fixed

**Comment by rettigl** [comment]
@zain-sohail I fixed most of the issues. Please comment on the remaining ones, then we can merge here.

File: sed/config/config_model.py [hunk 25]
```diff
@@ -0,0 +1,365 @@
...
+
+class StaticModel(BaseModel):
+    """Static configuration settings that shouldn't be changed by users."""
+
+    # flash specific settings
+    model_config = ConfigDict(extra="forbid")
+
+    stream_name_prefixes: Optional[dict] = None
+    stream_name_postfixes: Optional[dict] = None
+    beamtime_dir: Optional[dict] = None
```
[the remark this answers is not in the document]
Comment (zain-sohail) [reply sed/config/config_model.py:170]: It was just to have distinction between a user and say, an instrument scientist developing a loader. But I understand your point.

File: sed/config/config_model.py [hunk 26]
```diff
@@ -0,0 +1,365 @@
...
+    # flash specific settings
+    forward_fill_iterations: Optional[int] = None
+    ubid_offset: Optional[int] = None
+    split_sector_id_from_dld_time: Optional[bool] = None
+    sector_id_reserved_bits: Optional[int] = None
+    sector_delays: Optional[Sequence[float]] = None
+    daq: Optional[str] = None
+    num_trains: Optional[int] = None
+
+    # write validator for model so that x_column gets converted to columns: x
```
[the remark this answers is not in the document]
Comment (zain-sohail) [reply sed/config/config_model.py]: Yes compatiblity basically. But good point.

**Pull request merged**